xarray/coding/variables.py

```diff
@@ -566,11 +567,36 @@ def decode(self):
 
 class ObjectVLenStringCoder(VariableCoder):
     def encode(self):
-        return NotImplementedError
+        raise NotImplementedError
 
     def decode(self, variable: Variable, name: T_Name = None) -> Variable:
         if variable.dtype == object and variable.encoding.get("dtype", False) == str:
             variable = variable.astype(variable.encoding["dtype"])
             return variable
         else:
             return variable
+
+
+class EnumCoder(VariableCoder):
+    """Decode CF flag_* to python Enum"""
+
+    def encode(self, variable: Variable, name: T_Name = None) -> Variable:
+        raise NotImplementedError
+
+    def decode(self, variable: Variable, name: T_Name = None) -> Variable:
+        """From CF flag_* to python Enum"""
+        dims, data, attrs, encoding = unpack_for_decoding(variable)
+        if (
+            attrs.get("enum")
```

Comment:
There is no `enum` attribute on-disk for plain CF style enum. Testing for `flag_meaning`, `flag_values` should be sufficient. Problem: We would need to invent an enum name.

If we can decode from `flag_meaning`, `flag_values` to Python enum, it would be good to be able to roundtrip (implement `.encoding` Python enum to `flag_meaning`, `flag_values`).

Reply:
> But do we always want to decode CF flag_* into python enums ? Having this test also on "attrs['enum']" would avoid decoding them when this is not wanted.

No, only when `decode_enum=True`.

Reply:
> As for the encoding, I had previously a `EnumCoder::encode` method that turns an python Enum into CF flag_\*, but encoders are called before the `netCDF4_::open_store_variable` method. So within open_store_variable I would need to decode the flag_\* again and turn them into a netCDF4 enum, which defies the goal of having a python Enum, no ?

From my point of view, there are two things here. Encoding/decoding python enum <-> CF flag_\* and being able to read/write netCDF4 enum type. 

The current approach using attrs["enum"] to represent netCDF4 enum in xarray is straightforward and works very well.

We might just keep the CF stuff for another PR and get the netCDF4 enum backend feature in. @dcherian, do you have suggestions to move forward here?

Reply:
Agree with limiting to just the netCDF type, and skipping the `flag_*` stuff. 

Another thought I just had is whether we can use the `dtype` attribute instead. But we need two pieces of information: the `Enum` dict and the `dtype` of the values. Shall we stick the `Enum` dict in `dtype` and preserve the `dtype` of the array as the dtype on disk? Is there a usecase for encoding the values as a different dtype at all?

Reply:
I believe we can store the dtype within the enum with the `type` argument:  `encoding["dtype"] = Enum(e_name, e_dict, type=data.dtype.type)` but I don't know what are the caveats of this approach. 

> We might just keep the CF stuff for another PR and get the netCDF4 enum backend feature in.
> Agree with limiting to just the netCDF type, and skipping the flag_* stuff.

Alright, sorry for all these back and forths. I will try to focus on the netCDF Enum encoding/decoding.

Reply:
@bzah No need to apologize. :-) This stuff soon gets very complex and finding the best approach is based on multiple iterations.

I've thought a while on @dcherians comment above. We might be able to do something like this in `open_store_variable`:

```python
    if isinstance(var.datatype, netCDF4.EnumType):
        if not np.issubdtype(dtype, np.integer):
            raise TypeError("Only integer types can be used as enums")

        dtype = np.dtype(dtype, metadata={'enum': var.datatype.enum_dict,
                                          'enum_name': var.datatype.name})#

    encoding["dtype"] = dtype
```

This will use numpy's metadata to store the enum at the dtype. This is essentially the same as h5py is doing (without the `enum_name`). We could even provide that dtype to the variable itself. It's not that problematic to create a python enum from that dtype.

In `prepare_variable` we would just have to do the same thing analog to your current datatype extraction, but now from the variable.dtype

For that to work we would either need to implement EnumCoder.encoding to add back the metadata (it might have been stripped by processing) to the variable.dtype or fix `NonStringCoder` to do this (preferred, so we can reserve the EnumCoder for the flag_* stuff).

By using that approach we get the dtype and the enum (in dtype.metadata).